Use `amrex::SmallMatrix`

ax3l opened the pull request:
Replace current small matrix logic and math with the new `amrex::SmallMatrix` class.

ax3l commented on src/particles/diagnostics/EmittanceInvariants.cpp:
```diff
@@ -30,54 +30,53 @@ namespace impactx::diagnostics
         // permutation of the columns of Sigma with
         // a change of sign.
-        for (int i = 1; i < 7; i++) {
-            for (int j = 1; j < 7; j++) {
+        for (int i = 0; i < 6; i++) {
+            for (int j = 0; j < 6; j++) {
                 if (j % 2 != 0) {
-                   S1(i,j) = -Sigma(i,j+1); // if j is odd
+                   S1(i,j) = +Sigma(i,j-1); // if j is odd
                 }
                 else {
-                   S1(i,j) = +Sigma(i,j-1); // if j is even
+                   S1(i,j) = -Sigma(i,j+1); // if j is even
                 }
             }
         }
```
@cemitch99 @WeiqunZhang: shall we explicitly write `J` and multiply? That way, we do not need to worry about the fastest/slowest running index of the matrix.

cemitch99 commented on src/particles/elements/SoftSol.H:
```diff
@@ -518,30 +522,31 @@ namespace SoftSolenoidData
-            refpart.map(4,2) = R(4,2)*cs - R(2,2)*sn;
-            refpart.map(4,3) = R(4,3)*cs - R(2,3)*sn;
-            refpart.map(4,4) = R(4,4)*cs - R(2,4)*sn;
+            // note: map and R are zero-indexed
```
Hm.  This zero-indexing is a major concern.  This indices 1-6 for the transfer matrix are standard and widely used, and this is very likely to lead to confusion (especially for me).  I am strongly in favor of using an object indexed 1-6.

WeiqunZhang replied:
We could add a function to SmallMatrix that can be used to access the element using 1-based index. Not sure what to call it. But it could be something like `s.at<1>(i,j)` or a more explicit name.

WeiqunZhang replied:
Or add an optional template parameter for starting index to SmallMatrix

ax3l replied:
Fully agreed, thanks!

ax3l replied:
done

ax3l commented on src/particles/elements/SoftSol.H:
```diff
@@ -244,18 +244,21 @@ namespace SoftSolenoidData
-                  + R(6,4)*py + R(6,5)*t + R(6,6)*pt;
+            // note: R is zero-indexed
+            // clang-format off
+            xout =  R(0, 0) * x  + R(0, 1) * px + R(0, 2) * y
+                  + R(0, 3) * py + R(0, 4) * t  + R(0, 5) * pt;
+            pxout = R(1, 0) * x  + R(1, 1) * px + R(1, 2) * y
+                  + R(1, 3) * py + R(1, 4) * t  + R(1, 5) * pt;
+            yout =  R(2, 0) * x  + R(2, 1) * px + R(2, 2) * y
+                  + R(2, 3) * py + R(2, 4) * t  + R(2, 5) * pt;
```
@WeiqunZhang @cemitch99 can we express this as a matrix-vector multiplication?

cemitch99 replied:
Yes.  I did something similar in the draft of [PR#714](https://github.com/ECP-WarpX/impactx/pull/714), LinearMap.H, line 109.

ax3l replied:
I tried this but it failed the analysis so far:
```c++
amrex::SmallVector<amrex::ParticleReal, 6, 1> const v{x, px, y, py, t, pt};
auto const out = R * v;

// assign updated values
x  = out[0];
px = out[1];
y  = out[2];
py = out[3];
t  = out[4];
pt = out[5];
```

ax3l replied:
Ah, `out[1]`...

ax3l replied:
Done.

WeiqunZhang commented:
https://github.com/AMReX-Codes/amrex/pull/4188

ax3l committed:
Use `amrex::SmallMatrix`

Replace current small matrix logic and math with the new
`amrex::SmallMatrix` class.

ax3l committed:
Matrix-Vector Multiplication

More readable dense operation.

ax3l committed:
Eigenemittance: Fix Zero-Init

Ensure zero init of all non-set values

cemitch99 reviewed (approved):
Looks great, thanks!  This will make life easier.
